File: archive/c/c-sharp/ConvexHull.cs
using System.Runtime.InteropServices;

if (
    args is not [var xInput, var yInput]
    || !TryParsePoints(xInput.AsSpan(), yInput.AsSpan(), out var points)
)
{
    return Usage();
}

points.Sort();

var hull = BuildHull(CollectionsMarshal.AsSpan(points), out int size);
for (int i = 0; i < size; i++)
    Console.WriteLine(hull[i]);

return 0;

static bool TryParsePoints(ReadOnlySpan<char> x, ReadOnlySpan<char> y, out List<Point> points)
{
    points = [];

    if (x.IsWhiteSpace() || y.IsWhiteSpace())
        return false;

    var list = new List<Point>();

    while (true)
    {
        if (!TryNext(ref x, out int xVal) || !TryNext(ref y, out int yVal))
            break;

        points.Add(new(xVal, yVal));
    }

    return x.IsEmpty && y.IsEmpty && points.Count >= 3;

    static bool TryNext(ref ReadOnlySpan<char> span, out int value)
    {
        int comma = span.IndexOf(',');
        var token = comma >= 0 ? span[..comma] : span;
        span = comma >= 0 ? span[(comma + 1)..] : [];
        return int.TryParse(token, out value);
    }
}

static Point[] BuildHull(Span<Point> points, out int size)
{
    int numPoints = points.Length;
    if (numPoints < 3)
    {
        size = numPoints;
        return points.ToArray();
    }

    var hull = new Point[numPoints * 2];
    int h = 0;

    for (int i = 0; i < numPoints; i++)
    {
        var p = points[i];

        while (h > 1 && Cross(hull[h - 2], hull[h - 1], p) <= 0)
            h--;

        hull[h++] = p;
    }

    int lower = h;

    for (int i = numPoints - 2; i >= 0; i--)
    {
        var p = points[i];

        while (h > lower && Cross(hull[h - 2], hull[h - 1], p) <= 0)
            h--;

        hull[h++] = p;
    }

    size = h - 1;
    return hull;

    static long Cross(Point o, Point a, Point b) =>
        (long)(a.X - o.X) * (b.Y - o.Y) - (long)(a.Y - o.Y) * (b.X - o.X);
}

static int Usage()
{
    Console.Error.WriteLine(
        """Usage: please provide at least 3 x and y coordinates as separate lists (e.g. "100, 440, 210")"""
    );
    return 1;
}

public readonly record struct Point(int X, int Y) : IComparable<Point>
{
    public int CompareTo(Point other) => X != other.X ? X.CompareTo(other.X) : Y.CompareTo(other.Y);

    public override string ToString() => $"({X}, {Y})";
}

File: archive/v/visual-basic/convex-hull.vb
Module ConvexHull

    Public Structure Point
        Implements IComparable(Of Point)

        Public ReadOnly X As Integer
        Public ReadOnly Y As Integer

        Public Sub New(x As Integer, y As Integer)
            Me.X = x
            Me.Y = y
        End Sub

        Public Function CompareTo(other As Point) As Integer _
        Implements IComparable(Of Point).CompareTo

            Dim cx = X.CompareTo(other.X)
            If cx <> 0 Then Return cx
            Return Y.CompareTo(other.Y)
        End Function

        Public Overrides Function ToString() As String
            Return $"({X}, {Y})"
        End Function
    End Structure

    Sub Main(args As String())

        If args.Length <> 2 Then
            ShowUsage()
            Return
        End If

        Dim xCoords = ParseList(args(0))
        Dim yCoords = ParseList(args(1))

        If xCoords Is Nothing OrElse yCoords Is Nothing OrElse
           xCoords.Count <> yCoords.Count OrElse xCoords.Count < 3 Then

            ShowUsage()
            Return
        End If

        Dim points As New List(Of Point)(xCoords.Count)

        For i = 0 To xCoords.Count - 1
            points.Add(New Point(xCoords(i), yCoords(i)))
        Next

        Dim hull = BuildHull(points)

        For Each p In hull
            Console.WriteLine(p)
        Next

    End Sub

    Private Function BuildHull(points As List(Of Point)) As List(Of Point)

        Dim n = points.Count
        Dim hull As New List(Of Point)

        Dim start = 0
        For i = 1 To n - 1
            If points(i).CompareTo(points(start)) < 0 Then
                start = i
            End If
        Next

        Dim current = start

        Do
            hull.Add(points(current))

            Dim nextIdx = If(current + 1 = n, 0, current + 1)

            For i = 0 To n - 1
                If Cross(points(current), points(i), points(nextIdx)) > 0 Then
                    nextIdx = i
                End If
            Next

            current = nextIdx

        Loop While current <> start

        Return hull

    End Function

    Private Function Cross(o As Point, a As Point, b As Point) As Long
        Return CLng(a.X - o.X) * (b.Y - o.Y) -
               CLng(a.Y - o.Y) * (b.X - o.X)
    End Function

    Private Function ParseList(input As String) As List(Of Integer)

        If String.IsNullOrWhiteSpace(input) Then
            Return Nothing
        End If

        Dim result As New List(Of Integer)

        For Each part In input.Split(","c)

            Dim value As Integer

            If Not Integer.TryParse(part.Trim(), value) Then
                Return Nothing
            End If

            result.Add(value)

        Next

        If result.Count < 3 Then
            Return Nothing
        End If

        Return result

    End Function

    Private Sub ShowUsage()
        Console.Error.WriteLine(
            "Usage: please provide at least 3 x and y coordinates as separate lists (e.g. ""100, 440, 210"")"
        )
    End Sub

End Module